db/compaction/compaction_iterator.cc

```diff
@@ -178,6 +181,12 @@ void CompactionIterator::Next() {
       ikey_.user_key = current_key_.GetUserKey();
       validity_info_.SetValid(ValidContext::kMerge1);
     } else {
+      if (merge_until_status_.IsMergeInProgress()) {
+        // only merge operands were produced. reset the user key, since the
+        // batch produced by the merge operator should not shadow any keys
+        // coming after the merges
+        has_current_user_key_ = false;
```

Comment:
To be honest, I'm still trying to figure out whether the logic is correct. I will be helpful if we add some more comments to explain why this guarantees to output merge operands.

Reply:
Per https://github.com/facebook/rocksdb/blob/4d7cdcf260b790382f8a614305bb3be685c55f52/db/merge_helper.h#L75-L91, the original merge operands are accessible in `merge_out_iter_` after it returned `Status::MergeInProgress()`. At this particular line, `merge_out_iter_` is exhausted and the iterator is positioned at "the first entry not included in the merge process" (quote from that API doc). We know we need to output that even if it's in the same snapshot stripe because all we've outputted so far for this snapshot stripe is merges. I took this trick from elsewhere in the code. https://github.com/facebook/rocksdb/blob/4d7cdcf260b790382f8a614305bb3be685c55f52/db/compaction/compaction_iterator.cc#L923-L926

Reply:
Updated the comment. We have not necessarily finished processing the user key at this point; we have only finished processing the merge operands returned by the most recent call to `MergeUntil()`